Add missing babel helpers required for es7.decorators and es6.modules

pbastowski opened the pull request:
Added missing babel helpers required for es7.decorators and es6.modules (see also issue https://github.com/meteor/meteor/issues/5246).

The `ecmascript` Meteor 1.2 package supports custom whitelisting of various Babel transformers through the use of the `.babelrc` file, placed in the project root directory. This is standard Babel functionality, please read https://babeljs.io/docs/usage/babelrc/.

However, some of the transformers require in-browser runtime support through extra functions called Babel helpers, which are defined on the `babelHelpers` object in the `babel-runtime.js` file of the `babel-runtime` package. 

**ES7 decorators** and **ES6 modules** require the following helper functions, which are not included as standard: `createDecoratedClass`, `defineDecoratedPropertyDescriptor` and `interopRequireDefault`. This PR adds these helper functions to the `babel-runtime` package.

`.babelrc` must be formatted as in the example below.
```json
{
    "whitelist": [
        "es7.decorators",
        "es7.classProperties",
        "es7.exportExtensions",
        "es7.comprehensions",
        "es6.modules"
    ]
}
````

rclai commented:
Huh? Modules? Is this all it would take to get imports to work in Meteor?

pbastowski commented:
No it isn't :) But this at least transpiles the "import" keyword to an equivalent require call. You still need to supply the require library of some kind. I have my own `pbastowski:require` that does the job for my apps.

rclai commented:
Oh I see.

pbastowski commented:
That last commit's title was meant to say "Fixed bad createDecoratedClass function definition".

Mistereo commented:
Maybe it's better to use `.babelrc` (https://babeljs.io/docs/usage/babelrc/) instead of `babel.json`?

pbastowski commented:
Interesting that you mention this, as I also read this same page earlier today and was thinking about updating this PR. 

So, I think yes is the correct answer.

pbastowski commented:
While I was preparing the updated PR, which renames `babel.json` to `.babelrc`, I found something quite exciting. `.babelrc` is already implemented and provides the exact same functionality as my `babel.json` does!

So, no need to implement `babel.json`. I removed that functionality from this PR completely.

So, the only thing that was missing to enable the support for ES7 decorators and `import .... from` were the three babelHelpers functions that I added to `babel-runtime.js`. These were apparently removed on purpose (see README.md in the babel-runtime package), because they were described in Babel documentation as "experimental". Experimental as they may be, they are what we use in Angular2, the angular2-now library and TypeScript implements them too. 

The point is, we need ES7 decorators as part of Babel and this PR adds the babelHelpers functions that are needed to support them.



So, this PR now consists only of the additions to the `babel-runtime.js` file.

pbastowski commented:
OK, this is the final nail in the coffin for this PR, which I will close after some of you good people help me verify my findings. below:

- As I mentioned in my previous posts above, `.babelrc` allows us to whitelist extra Babel transformers
- Some transformers require specific helper functions to be added to the `babalHelpers` object
- I just discovered that the `babelHelpers` object is visible on `window`, so, we can add to it whatever functions we wish and desire.

So, create a file `client/lib/babel-helpers.js` in your project and add the following to it:
````javascript
// Provides support for es7.decorators
window.babelHelpers.defineDecoratedPropertyDescriptor = function (target, key, descriptors) {
    var _descriptor = descriptors[key];
    if (!_descriptor) return;
    var descriptor = {};

    for (var _key in _descriptor) descriptor[_key] = _descriptor[_key];

    descriptor.value = descriptor.initializer ? descriptor.initializer.call(target) : undefined;
    Object.defineProperty(target, key, descriptor);
};

// Provides support for es7.decorators
window.babelHelpers.createDecoratedClass = (function () {
    function defineProperties(target, descriptors, initializers) {
        for (var i = 0; i < descriptors.length; i++) {
            var descriptor = descriptors[i];
            var decorators = descriptor.decorators;
            var key = descriptor.key;
            delete descriptor.key;
            delete descriptor.decorators;
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ('value' in descriptor || descriptor.initializer) descriptor.writable = true;
            if (decorators) {
                for (var f = 0; f < decorators.length; f++) {
                    var decorator = decorators[f];
                    if (typeof decorator === 'function') { descriptor = decorator(target, key, descriptor) || descriptor; } else { throw new TypeError('The decorator for method ' + descriptor.key + ' is of the invalid type ' + typeof decorator); }
                }
                if (descriptor.initializer !== undefined) {
                    initializers[key] = descriptor;
                    continue;
                }
            }
            Object.defineProperty(target, key, descriptor);
        }
    }

    return function (Constructor, protoProps, staticProps, protoInitializers, staticInitializers) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps, protoInitializers);
        if (staticProps) defineProperties(Constructor, staticProps, staticInitializers);
        return Constructor;
    };
})();

window.babelHelpers.interopRequireDefault = function (obj) {
    return obj && obj.__esModule ? obj : {
        "default": obj
    };
};
````

Now, create the file `.babelrc` in the root of your project and add the following to it:
```javascript
{
    "whitelist": [
        "es7.decorators",
        "es7.classProperties",
        "es7.exportExtensions",
        "es7.comprehensions",
        "es6.modules"
    ]
}
```

Now, your application that uses the `ecmascript` package and uses ES7 decorators, possibly because it uses the `pbastowski:angular2-now` package, should process both decorators and `import .... from` statements as expected and previously accomplished with `grigio:babel` and/or my very own `pbastowski:angular-babel` packages.

Could someone please let me know if the above worked for them?

Thanks.

pbastowski commented:
Closed, because it is not required as per my explanation above.

stubailo commented:
OK, sounds good! I think we'll be improving our ES2015+ support over time anyway, hopefully it meets your needs!

pbastowski commented:
@stubailo  I create a package `pbastowski:ecmascript-extras`, which adds the features I mentioned above on top of the existing `ecmascript` package.
